src/System.Net.Sockets/src/System/Net/Sockets/TCPClient.Unix.cs

```diff
@@ -0,0 +1,424 @@
+using System.Runtime.ExceptionServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Sockets
+{
+    partial class TcpClient
+    {
+        // The Unix implementation is separate from the Windows implementation due to a fundamental
+        // difference in socket support on the two platforms: on Unix once a socket has been used
+        // in a failed connect, its state is undefined and it can no longer be used (Windows supports
+        // this).  This then means that the instance Socket.Connect methods that take either multiple
+        // addresses or a string host name that could be mapped to multiple addresses via DNS don't
+        // work, because we can't try to individually connect to a second address after failing the
+        // first.  That then impacts TcpClient, which exposes not only such Connect methods that
+        // just delegate to the underlying Socket, but also the Socket itself.
+        //
+        // To address the most common cases, where the Socket isn't actually accessed directly by
+        // a consumer of TcpClient, whereas on Windows TcpClient creates the socket during construction,
+        // on Unix we create the socket on-demand, either when it's explicitly asked for via the Client
+        // property or when the TcpClient.Connect methods are called.  In the former case, we have no
+        // choice but to create a Socket, after which point we'll throw PlatformNotSupportedException
+        // on any subsequent attempts to use Connect with multiple addresses.  In the latter case, though,
+        // we can create a new Socket for each address we try, and only store as the "real" Client socket
+        // the one that successfully connected.
+        //
+        // However, TcpClient also exposes some properties for common configuration of the Socket, and
+        // if we simply forwarded those through Client, we'd frequently end up forcing the Socket's
+        // creation before Connect is used, thwarting this whole scheme.  To address that, we maintain
+        // shadow values for the relevant state.  When the properties are read, we read them initially
+        // from a temporary socket created just for the purpose of getting the system's default. When
+        // the properties are written, we store their values into these shadow properties instead of into
+        // an actual socket (though we do also create a temporary socket and store the property onto it,
+        // enabling errors to be caught earlier).  This is a relatively expensive, but it enables
+        // TcpClient to be used in its recommended fashion and with its most popular APIs.
+
+        // Shadow values for storing data set through TcpClient's properties.
+        // We use a separate bool field to store whether the value has been set.
+        // We don't use nullables, due to one of the properties being a refernece type.
+
+        private int _exclusiveAddressUse;
```

Comment:
Might it make sense to bundle these properties in a lazily-initialized class? It would make this a bit more pay-for-play.